test.py

```diff
@@ -161,17 +270,127 @@
+                [
+                    np.mean(
+                        [
+                            1.0
+                            - diceloss(
+                                torch.where(
+                                    total_scores[i] < thresh,
+                                    torch.zeros_like(total_scores[i]),
+                                    torch.ones_like(total_scores[i]),
+                                ),
+                                total_target[i],
+                            ).item()
+                            for i in tqdm(
+                                range(total_target.shape[0]),
+                                total=total_target.shape[0],
+                                leave=False,
+                            )
+                        ]
+                    )
+                    for thresh in tqdm(
+                        thresholds,
+                        total=len(thresholds),
+                        leave=False,
+                        desc="optimizing thresholds",
+                    )
+                ]
+            )
+            print(
+                f"Best dice score was {np.max(dicescores)*100.0:.2f} @ {thresholds[np.argmax(dicescores)]}"
+            )
+            print(
+                f"Overall scores @ threshold levels: {[*zip(thresholds, dicescores)]}"
+            )
+            threshold = thresholds[np.argmax(dicescores)]
+        else:
+            threshold = 0.5
+        total_pred = torch.where(
+            total_scores < threshold,
+            torch.zeros_like(total_scores),
+            torch.ones_like(total_scores),
+        )
+        losses = [
+            diceloss(total_scores[i], total_target[i]).item()
+            for i in tqdm(
+                range(total_target.shape[0]), total=total_target.shape[0], leave=False
+            )
+        ]
+        scores = [
+            1.0 - diceloss(total_pred[i], total_target[i]).item()
+            for i in tqdm(
+                range(total_target.shape[0]), total=total_target.shape[0], leave=False
+            )
+        ]
+        iou = smp.utils.metrics.IoU()(total_pred, total_target)
+        # fscore = smp.utils.metrics.Fscore()(total_pred, total_target)
+        print(f"Dice Loss: {np.mean(losses)*100.0:.2f}%")
+        print(f"Dice Score @ threshold {threshold}: {np.mean(scores)*100.0:.2f}%")
+        print(f"IoU: {iou*100.0:.2f}%")
+        # print(f"Fscore: {fscore*100.0:.2f}%")
+
+        interesting_idcs = np.argsort(
+            np.asarray(scores)
+        )  # could be more efficient using argpartition but this is simple and doesn't really matter for this number of array entries
+        interesting_idcs = np.concatenate([interesting_idcs[:5], interesting_idcs[-5:]])
+
+        plot_imgs(
+            imgs[interesting_idcs],
+            total_target[interesting_idcs],
+            total_pred[interesting_idcs],
+            savefig="test.png",
+        )
+
+        # plt.clf()
```

Comment:
Should we remove the unused comments here? @NiWaRe

Reply:
Yes sure there're actually a ton of comments I still need to remove, will do so asap.